### ISSUE TITLE
Embedding of external functions

### ISSUE THREAD

**Issue by vincentvanhees** [wadpac/GGIR#241]
The plan is to allow for specification of an external function to be applied to the raw accelerometer data, and to integrate the output of that function in the GGIR output. For example, an external algorithm for sleep detection, step detection, cycling detection, etc. The advantages include:
- Avoids embedding all these algorithms in GGIR itself.
- Gives freedom to developer to develope their code and models independent of GGIR.
- Open to dependencies in other programming languages, e.g. Matlab or Python.
- Allows for new features to be developed as external function first, and if they are well tested we can include them eventually as a GGIR function if external developers agrees.

**Plan for now:**
- [x] Create development branch.
- [x] Add function object as extra argument to g.shell.GGIR, g.part1, g.getmeta.
- [x] Make g.getmeta apply this function argument to blocks of raw accelerometer data.
- [x] Define clear API requirements for this function
- [x] Adapt function such that it can include a loaded model, by which we do not have to load the model with every new block of data. So, rather than providing a function, we expect the user to provide a function and a separately loaded model, which is then fed into the function by GGIR.
- [x] Include functionality to resample the input data if external function requires this.
- [x] Include functionality to resample the output data if external function does not do this.
- [x] Make getmeta truncate the data if necessary to be convenient for external function, e.g. integer number of window sizes.
- [x] Tidy up g.getmeta by moving all the new code to a separate function + write unit tests.
- [x] Update specific documention about myfun object (g.getmeta, g.part1, g.shell.GGIR)
- [x] Make g.getmeta add the external function output to the part1 milestone data.
- [x] Integrate external function produced variable names, such that features can be used in parts 2, 3, 4 and 5
- [x] Add examples of how to use this with Python code (reticulate package).
- [x] Create vignette with overview of examples of R function (count calculation) and Python function (dominant frequency calculation).
- [x] Address failing OSx-specific travis build, related to using python in Rmarkdown.
- [x] Add option to pass (numeric) timestamps to external function together with data object.
- [x] Add quality checks on myfun object, and on output. Maybe worth defining an explicit class for it.
- [x] Enable interpolation, e.g. 30 second classification to 5 second output
- [x] Embed output in part2 report (events)
- [x] Expand vignette tutorial with more information about output, update date.
- [x] Put example vignette pdf in repo

**Embed sleep/nonwear classification output LRAP project in part 2 and 4 reports:**
- [x] Embed output in part 4 report (sleep): Write part 3 function that extracts sleep from part 1 and formats it in a way that part 4 can handle.
- [ ] Fuse non-wear with conventional non-wear in part 2. If shorter than 1 hour rely on new classification, otherwise use old one.

**Comment by vincentvanhees** [comment]
Current thoughts on S3 class object for the input data:
```
rawacc <- list(acc = data.frame(accx=x, accy=y, accz=z, original_sample_rate = 100, current_sample_rate = 50, unit="g")
class(rawacc) <- "rawacc"
```

- Object has no timestamps, because I assuming that algorithms do not need timestamps.
- I am including both the original and the current sample rate of the data, because algorithms may be sample rate specific which requires resampling. However, to preserve knowledge about the original sample rate we also store this information.
- Unit, because algorithms may be measurement unit specific.

**Comment by vincentvanhees** [comment]
An alternative approach could be to define a class around the external function, rather than around the data.

mytool = list(FUN, expected_sample_rate=50, expected_unit="g", expected_columns="xyz"
class(mytool) = "external function")

In that way it is clear for the external function developer what to provide.

**Comment by m-patterson** [comment]
Thanks very much for setting up this framework Vincent.

However, I am still having trouble getting my function set up in the right way to work properly. I've got a version of 'issue241_ExtrnlFuncEmbedding', and I have my external function to count steps, but I'm not having much luck getting it working in the framework.

I want it to work on ws2 sized chunks of raw acc data and it can use any sampling rate, it just needs to know the sample rate so it can adjust thresholds accordingly.

I tried adding it to the input list in my entry function, like this:
myfun = c(findsteps_peakdet())

and it requires the data chunk as an input and the sampling rate, but I don't see in the code how I can set that up properly, or make it in the format in which the structure you have set up requires. 

I'm getting an error that the data is not available inside my function.

Also, should I be saving my function in the GGIR directory with all the other .R files? (GGIR/R)

Any guidance on setting up my function properly and calling it in entry function would be much appreciated. Thanks, Matt

**Comment by m-patterson** [comment]
Hi Vincent, thanks for all the work and documentation on this. I'm a bit stuck in the 'applyExtFunction.R' script, as there is a call to a 'resample' function. Does this function come from  library, or should it be included in the .R files with GGIR? IT doesn't seem to be in the issue241_ExtrnlFuncEmbedding branch?

IT is inside the resampleAcc function, which is defined inside applyExtFunction.R:
![image](https://user-images.githubusercontent.com/16383946/74472016-38d45f80-4e6f-11ea-8cc1-b3044831bd79.png)

**Comment by vincentvanhees** [comment]
Hi Matt,
resample is a c++ function from the src directory, which needs to be compiled with (only if you are working with the code in development mode):
```
library("Rcpp")
pathR = "/home/GGIR"
sourceCpp(paste0(pathR,"/src/numUnpack.cpp"))
sourceCpp(paste0(pathR,"/src/resample.cpp"))
```

I am currently trying to fix some issues in this part of the code. so please hold on. I am hopeful that I can commit my changes in the upcoming hour or so. This relates robustness to incorret use of myfun$timestamp value.

**Comment by m-patterson** [comment]
sounds good, thanks!

**Comment by vincentvanhees** [comment]
I am done for the day, the above commit was what I wanted to do. The counts example and the dominant frequency example still seem to work.

**Comment by m-patterson** [comment]
Thanks for those updates Vincent, it seems to be working well now. For, the step count algorithm, the user will want to have the number of steps taken each day as well as the total number of steps taken over the entire recording? Do you think you will set this up to auto generate in the output, or will there be some sort of myfun setting to indicate how the final output should be shown?

Great work on this tool, it is a very powerful way to quickly apply different algorithms to long term activity recordings!

**Comment by vincentvanhees** [comment]
Thanks Matt for letting me know.
Yes, I think some input will be needed to make it generically useful.

Possible types of features:
- event count (e.g. steps, falls, ...)
- labels (e.g. activity types)
- scalar (e.g. likelihood that someone fell, energy expenditure estimate, ...)
- multivariate scalars (e.g. some scalar calculated per accelerometer axis, frequency spectrum, ...)

Possible types of aggregating per day and per recording:
- average
- sum

Possible desired ways of reporting:
- Part 2
- Use as alternative sleep detection in parts 3 and 4
- Part 5

Next, the myfun object may then need to carry a choice for each of these three categories. For example, to be named: outputtype = "eventcount", how2aggregate4report = "sum", and GGIRreport = 2.
Can you think of any additional types or decisions that need to be accounted for to be future proof?

**Comment by m-patterson** [comment]
I think that is a pretty thorough list of how algorithm creators will want to categorize outputs.

There might be a need to only include results from either day or night or else to separate day results and night results. Maybe this would take more time to implement, so could be something to do down the road. For the steps algorithm I don't need it.

**Comment by vincentvanhees** [comment]
Can you send me an example milestone file with the step count in it?
I could use that to develop the automated incorporation of total steps per day in the csv report.

**Comment by vincentvanhees** [comment]
Closing this issue as it seems to be working for a variety of use cases.

Most recent addition is the use of an external function for sleep detection and embedding in all sleep and part 5 reports. I have added a pdf copy of the tutorial to the repo https://github.com/wadpac/GGIR/tree/GGIR2.0/pdf_vignettes_for_github_only

**Inclusion in new release:**
I merged all progress with branch GGIR2.0, which I aim to merge with the master branch in the upcoming weeks and that will be release 2.0-0 (GitHub only).  Next, I will encourage the GGIR e-mail list (google group) to start testing it, followed by 2.1-0 a couple of weeks later on GitHUb and on CRAN.  

New ideas for improvement can be collected in a new issue, for now I think it is good enough.
